chrome_extension/background.js

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+// TODO: Generate an ECDSA key pair, and use the private key for authentication and the public key
+// as Oak label.
+//
+// See https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/sign
+
+// Ids of tabs that have entered Oak mode.
+const enabled_tabs = new Set();
+
+// Intercept outgoing requests, and block them if they are coming from a tab that has entered Oak
+// mode.
+//
+// TODO: This only intercepts new WebSocket connections, but not individual messages on previously
+// established connections. e.g. https://www.websocket.org/echo.html.
+chrome.webRequest.onBeforeRequest.addListener(
```

Comment:
Does this also catch content loading in child IFrames? Does this catch opening of new windows/tabs?

Reply:
iframes: yes
new windows: no (I was able to exfiltrate data via 

```js
window.open('https://google.com/?q=123');
```

@juliettepretot do you think there is a any way of catching this?

Reply:
There's a CSP working draft to prevent this, but so far it's not implemented in any browser. 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/navigate-to

`chrome.webNavigation.onBeforeNavigate` might be able to stop it, but I haven't looked into it: https://developer.chrome.com/extensions/webNavigation#event-onBeforeNavigate

Reply:
On a higher level I'm a bit unsure how practical the approach of treating all client code as untrusted is. In practice that means pretty heavy restrictions. Any oak website couldn't include external links, as it'd risk the same kind of data leak.

Reply:
And even if those are successfully imposed, data can still be leaked by other evil chrome extensions reading the Oak website's content.

Reply:
yeah good point re: external links.

I think eventually we would need something more drastic, like incognito mode, in which even links that you open from incognito are themselves opened in more incognito tabs. But this is a good starting point I think, despite all the assumptions.